src/libsyntax/parse/parser.rs

```diff
@@ -4763,6 +4763,8 @@ impl<'a> Parser<'a> {
                     err.span_label(sp, &"missing `fn`");
                 }
                 return Err(err);
+            } else if let Err(bang_err) = bang_err {
+                return Err(bang_err);
```

Comment:
I feel that if this code used explicit match, mistake wouldn't have been made. Something like

```
 match (err, bang_err) {
    (Err(_), Err(_)) => ...
    (Err(_), _) => ...
    ...
}

```